azure/services/groups/groups.go

```diff
@@ -67,9 +97,34 @@ func (s *Service) Reconcile(ctx context.Context) error {
 
 	err := async.CreateResource(ctx, s.Scope, s.client, groupSpec, serviceName)
 	s.Scope.UpdatePutStatus(infrav1.ResourceGroupReadyCondition, serviceName, err)
+
+	if err != nil {
+		return err
+	}
+
+	err = s.ReconcileTags(ctx)
```

Comment:
Instead of calling ReconcileTags here directly in the group service, what do you think about adding a step in the azurecluster reconciler to call tagsSvc.Reconcile() just like we do already in azuremachine_reconciler for VMs? That way, if we later want to expand the tag updates to other resource types (eg. VNet) in AzureCluster we can just append it to TagSpecs.

Reply:
Makes sense! I'll do that 👍

Reply:
What about Azure Managed Control Plane reconciler? Should I add to that too? I have already started to handle that too in this PR. If we are looking for a separate issue and PR on that then I can remove those changes from this PR

Reply:
Also, if we move tag reconciliation of resources to the Azure Cluster Reconciler, how do we handle the part about checking if the resource is managed or not? The tag service reconciliation simply loops over all the tag specs and does reconciliation and works only with Kubernetes level data, so if some resources are not managed at some point (owned value tag removed) then those have to be skipped in the tag reconciliation 🤔 while others have to reconcile

Reply:
At Azure Cluster Reconciler level, we don't have full access to the services and azure API clients of the resources, so finding if a resource is managed or not from that level is going to be a bit tricky. I was thinking of two places to put the is-managed check code - either in the reconcileTags or in tag specs in each tag spec, but again, both are tricky and can bloat up pretty quickly as we want managed tags for more resources

Reply:
https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1721/files#r715302634

Reply:
@karuppiah7890 here are two ways we could workaround this:

1) Do something like https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1684/files#diff-b74c6acb4629abe58f8286da9939563cc4404e09fe9966eb61dbde9b87274189R113 for the RG (store management state at the beginning of the reconcile and use it later to build the tag spec)
2) since we're already getting the existing tags in the tag service, and this is what we use to determine if a resource is managed, we could add a condition for only reconciling tags when the set of existing tags contains the cluster name owned tag. This would also apply to other resources using the tag service (just VMs right now) so we would need to either a) add owned tags to VMs, which we're not doing currently or b) add some bool field to the tagSpec to specify if we should check resource management state and set it to true for groups and false for VMs.

What do you think?

Reply:
I was also thinking about doing the check in the tag service the only odd one out was VMs case, and yeah we could also use the boolean I thought about that too. But if VM is not an exception, that would make things better, just that it would mean touching other parts of the code not related to RG, which could probably done in a separate PR

Reply:
About storing state at the start - I have to check it out, didn't think about that

Reply:
Great ideas, I'll check them out and get back with any questions or thoughts, maybe along with some code

Reply:
> add owned tags to VMs, which we're not doing currently

I take this back, looks like we are already adding the same tags to VMs so it shouldn't be an exception: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/azure/services/virtualmachines/virtualmachines.go#L145

Reply:
> I was also thinking about doing the check in the tag service

I was thinking about pulling tags by pulling the whole resource 🙈 I just realized what you meant about tags service pulling tag using `get-at-scope` Tag API which I didn't think about using

By the way, I'm planning to use the update-at-scope Tag API to avoid race conditions and use get-at-scope Tag API to just check if the resource group is managed or not. The only advantage of doing the API call and not using cache would be - updated data - as the it's possible that resource group can get created with owned tag but then the resource group can lose the owned tag value when reconciling the tags, say some external system deleted it. But yeah, it's not exactly a usual thing so the cache will still be helpful in most cases as long as no one is meddling with CAPZ's owned tag value and corresponding tag name